module/dkg/client.go

```diff
@@ -0,0 +1,246 @@
+	data, err := json.Marshal(msg)
+	if err != nil {
+		return fmt.Errorf("could not marshal DKG messages struct: %v", err)
+	}
+
+	// add dkg message json encoded string to tx args
+	err = tx.AddArgument(cadence.NewString(string(data)))
+	if err != nil {
+		return fmt.Errorf("could not add whiteboard dkg message to transaction: %w", err)
+	}
+
+	// sign envelope using account signer
+	err = tx.SignEnvelope(c.accountAddress, int(c.accountKeyIndex), c.signer)
+	if err != nil {
+		return fmt.Errorf("could not sign transaction: %w", err)
+	}
+
+	// submit signed transaction to node
+	txID, err := c.submitTx(ctx, tx)
+	if err != nil {
+		return fmt.Errorf("failed to submit transaction: %w", err)
+	}
+
+	// wait for transaction to be sealed
+	result := &sdk.TransactionResult{Status: sdk.TransactionStatusUnknown}
+	for result.Status != sdk.TransactionStatusSealed {
+		result, err = c.flowClient.GetTransactionResult(ctx, txID)
+		if err != nil {
+			return fmt.Errorf("could not get transaction result: %w", err)
+		}
+
+		// if the transaction has expired we skip waiting for seal
+		if result.Status == sdk.TransactionStatusExpired {
+			return fmt.Errorf("broadcast dkg message transaction has expired")
+		}
+
+		// wait 1 second before trying again.
+		time.Sleep(time.Second)
+	}
+
+	return nil
+}
+
+// ReadBroadcast reads the broadcast messages from the smart contract.
+// Messages are returned in the order in which they were broadcast (received
+// and stored in the smart contract)
+func (c *Client) ReadBroadcast(fromIndex uint, referenceBlock flow.Identifier) ([]model.DKGMessage, error) {
+
+	ctx := context.Background()
+
+	// construct read latest broadcast messages transaction
+	template := templates.GenerateGetDKGLatestWhiteBoardMessagesScript(c.env)
+	value, err := c.flowClient.ExecuteScriptAtBlockID(ctx, sdk.Identifier(referenceBlock), template, []cadence.Value{cadence.NewInt(int(fromIndex))})
+	if err != nil {
+		return nil, fmt.Errorf("could not execute read broadcast script: %w", err)
+	}
+	values := value.(cadence.Array).Values
+
+	// unpack return from contract to `model.DKGMessage`
+	messages := make([]model.DKGMessage, 0)
+	for _, val := range values {
+
+		content := val.(cadence.Struct).Fields[1]
+		jsonString, err := strconv.Unquote(content.String())
+		if err != nil {
+			return nil, fmt.Errorf("could not unquote json string: %w", err)
+		}
+
+		var flowMsg model.DKGMessage
+		err = json.Unmarshal([]byte(jsonString), &flowMsg)
+		if err != nil {
+			return nil, fmt.Errorf("could not unmarshal dkg message: %w", err)
+		}
+		messages = append(messages, flowMsg)
+	}
+
+	return messages, nil
+}
+
+// SubmitResult submits the final public result of the DKG protocol. This
+// represents the group public key and the node's local computation of the
+// public keys for each DKG participant. Serialized pub keys are encoded as hex.
+func (c *Client) SubmitResult(groupPublicKey crypto.PublicKey, publicKeys []crypto.PublicKey) error {
```

Comment:
They are ordered on the crypto library level, and the order should not be changed afterwards. Good flag.

Reply:
Btw, was the contract implemented to see the final submission as an array of strings, or as one long string? 
Fyi, the contract doesn't need to know the details of the array and it can consider the final submission as one string.

Reply:
The contract sees [an array of strings](https://github.com/onflow/flow-core-contracts/blob/feature/epochs/contracts/epochs/FlowDKG.cdc#L33)

Reply:
Great. 
A minor optimization: We can use the first string as the group public key , as done in this PR. The reason is that the contract compares the final submissions going through keys starting from index `0`. The group public key is the one usually being different when final submissions are different. This way the comparison fails early.